Scheduler.lib.nut

```diff
@@ -0,0 +1,430 @@
+
+        if (t < now.time) t = now.time;
+
+        local newJob = Job(this, {
+            "type": JOB_TYPE_AT,
+            "id": _nextId++,
+            "cb": cb,
+            "args": vargv
+        });
+
+        if (_env == ENVIRONMENT_AGENT) {
+            newJob.sec = t;
+            newJob.subSec = 0.0;
+        } else {
+            newJob.sec = math.floor(nowMillis).tointeger() + (t - now.time);
+            newJob.subSec = nowMillis - math.floor(nowMillis).tointeger();
+        }
+
+        _addJob(newJob);
+
+        return newJob;
+    }
+
+    // Start a new timer to trigger repeatedly at a specific interval
+    //
+    // Parameters:
+    //     int (float)     the time between executions of the timer
+    //     cb  (function)    the function to run when the timer fires
+    // Return: (integer) the id number of the timer (can be used to cancel the timer)
+    function repeat(int, cb, ...) {
+        try {
+            local now = _getTime();
+            vargv.insert(0, null);
+
+            if (int < 0) int = 0;
+
+            local newJob = Job(this, {
+                "type": JOB_TYPE_REPEAT,
+                "id": _nextId++,
+                "sec": _getSec(int, now),
+                "subSec": _getSubSec(int, now),
+                "repeat": int,
+                "cb": cb,
+                "args": vargv
+            });
+            _addJob(newJob);
+
+            return newJob;
+        } catch(e) {
+            server.error(e);
+            throw e;
+        }
+    }
+
+    // Start a new timer to trigger repeatedly at a specific interval, starting at a specific time
+    //
+    // Parameters:
+    //     t   (float)     the time for the first execution of the timer
+    //     int (float)     the time between executions of the timer
+    //     cb  (function)    the function to run when the timer fires
+    // Return: (integer) the id number of the timer (can be used to cancel the timer)
+    function repeatFrom(t, int, cb, ...) {
+        local now = date();
+        local nowMillis = null;
+        if (_env != ENVIRONMENT_AGENT) nowMillis = hardware.millis() / 1000.0;
+
+        vargv.insert(0, null);
+
+        if (int < 0) int = 0;
+        if (t < now.time) t = now.time;
+
+        local newJob = Job(this, {
+            "type": JOB_TYPE_REPEAT_FROM,
+            "id": _nextId++,
+            "repeat": int,
+            "cb": cb,
+            "args": vargv
+        });
+
+        if (_env == ENVIRONMENT_AGENT) {
+            newJob.sec = t;
+            newJob.subSec = 0.0;
+        } else {
+            newJob.sec = math.floor(nowMillis).tointeger() + (t - now.time);
+            newJob.subSec = nowMillis - math.floor(nowMillis).tointeger();
+        }
+
+        _addJob(newJob);
+
+        return newJob;
+    }
+
+    function tzoffset(offset = null) {
+        // Store and retrieve the tzoffset from the global scope
+        if (!("timer_tzoffset" in ::getroottable())) ::timer_tzoffset <- 0;
```

Comment:
why do you think it should be a global variable? can it be a property of the instance/class here?

Reply:
That was a remnant of the old version. I can make it a class instance.

Reply:
I've removed the method because it isn't required anymore now that we're not doing date strings.